lib/graphql/cache/marshal.rb

```diff
@@ -50,10 +50,24 @@ def read(config, force: false, &block)
       # @param config [Hash] The middleware resolution config hash
       def write(config)
         resolved = yield
+
         document = Deconstructor[resolved].perform
 
-        cache.write(key, document, expires_in: expiry(config))
-        resolved
+        with_resolved_document(document) do |resolved_document|
+          cache.write(key, resolved_document, expires_in: expiry(config))
+
+          resolved
+        end
+      end
+
+      # @private
+      def with_resolved_document(document)
+        if document.is_a?(GraphQL::Execution::Lazy) ||
+           self.class.const_defined?('Promise') && document.is_a?(Promise)
```

Comment:
I would recommend refactoring this conditional's logic into another private method.  Something like

```ruby
def document_is_lazy(document)
  document.is_a?(GraphQL::Execution::Lazy) || self.class.const_defined?('Promise') && document.is_a?(Promise)
end
```
That way the `with_resolved_document` method is a bit more clear in intention.  Also, what are your thoughts on using `document.class.name` instead of the combined conditional with `const_defined?` and `is_a?`?

Reply:
Good call. And yeah better to do `document.class.name` rather than check if a const is defined, both here and in the deconstructor.